group search rest api search get

koevskinikola opened the pull request:
## Description

<!-- Describe the goal and purpose of this PR. -->
Add Group REST API endpoints for:
- get Group (HTTP GET)
- search Groups (HTTP POST)

The PR also implements a separate `GroupQueryController` that will encapuslate Group-related endpoints that query the ES/OS backend.

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)

## Related issues

related to #22387

remcowesterhoud commented on zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/SearchQueryRequestMapper.java:
```diff
@@ -129,6 +131,20 @@ public static Either<ProblemDetail, RoleQuery> toRoleQuery(final RoleSearchQuery
 
+  public static Either<ProblemDetail, GroupQuery> toGroupQuery(
+      final GroupSearchQueryRequest request) {
+    if (request == null) {
+      return Either.right(SearchQueryBuilders.groupSearchQuery().build());
+    }
```
❓ So if there is no request body applied, we will just return all the groups that exist?

remcowesterhoud commented on zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/controller/usermanagement/GroupQueryControllerTest.java:
```diff
@@ -95,4 +100,111 @@ void shouldReturnNotFoundOnGetNonExistingGroup() {
+  }
+
+  @Test
+  void shouldSortAndPaginateSearchResult() {
```
❌ This test doesn't actually assert the result of the request. Shouldn't we make sure we limit the page to 2 results, verify we only get 2 of the group entities and that they are ordered correctly?

koevskinikola replied:
Yup, that makes sense, thanks Remco!

remcowesterhoud commented on zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/controller/usermanagement/GroupQueryControllerTest.java:
```diff
@@ -95,4 +100,111 @@ void shouldReturnNotFoundOnGetNonExistingGroup() {
+  }
+
+  @Test
+  void shouldSortAndPaginateSearchResult() {
```
❌ I'm missing test cases for the exceptional cases. What if the field in sort is invalid? What if the provided order is invald? What if I provide invalid values to page, such as a negateive limit?

koevskinikola replied:
Good point, I've added them.

remcowesterhoud reviewed (changes requested):
Thanks @koevskinikola! 🚀 I'm missing some test cases though.

koevskinikola commented:
@remcowesterhoud I added the missing tests.

remcowesterhoud commented on zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/controller/usermanagement/GroupQueryControllerTest.java:
```diff
@@ -193,18 +226,134 @@ void shouldSortAndPaginateSearchResult() {
-        .isOk();
+        .isOk()
+        .expectBody()
+        .json(EXPECTED_SEARCH_RESPONSE);
```
❓ If the limit is set to 2 wouldn't we only expect 2 of the groups in the result?

koevskinikola replied:
Something isn't workin as expected, I'll need to investigate some more.

koevskinikola replied:
@remcowesterhoud since we're mocking the `GroupService` the actual sort and filter parameters are not considered. Basically, with the test we confirm that the requests and responses are properly parsed and mapped.

We have separate `GroupService` tests that cover the sorting and filtering.

koevskinikola commented:
❌ I just noticed I'm missing the actual Group-specific filters in the search request. I need to add them.

remcowesterhoud commented on zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/controller/usermanagement/GroupQueryControllerTest.java:
```diff
@@ -193,18 +226,134 @@ void shouldSortAndPaginateSearchResult() {
                 .build());
   }
+
+  @Test
+  void shouldFailAnInvalidSearchQuery() {}
```
❌ I think you forgot to remove this one 🙂

remcowesterhoud commented on zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/controller/usermanagement/GroupQueryControllerTest.java:
```diff
@@ -193,18 +226,134 @@ void shouldSortAndPaginateSearchResult() {
+  void shouldFailAnInvalidSearchQuery() {}
+
+  @ParameterizedTest
+  @MethodSource("invalidGroupSearchQueries")
```
👍 I like the parameterization!

koevskinikola replied:
I can't take credit for it, it was already very well done in the `AuthorizationQueryControllerTest` 🙂

remcowesterhoud reviewed (approved):
@koevskinikola I have one open question and you msised to cleanup one test case. I'll pre-approve it :)

koevskinikola commented:
@remcowesterhoud I added the missing properties and cleaned up the tests. Do you want to do another review on the code, or the pre-approval is enough?

koevskinikola commented:
@remcowesterhoud to make faster progress and avoid making these branch PRs too stale, I merged this PR. Let me know if you have additional feedback, and I'll piggyback in on later Group-related REST API PRs.

remcowesterhoud commented:
> Do you want to do another review on the code, or the pre-approval is enough?

Nah, I'm good with it
